rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,277 @@
+// Copyright 2020 DCS Corporation, All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// DISTRIBUTION A. Approved for public release; distribution unlimited.
+// OPSEC #4584.
+//
+// Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
+// Part 252.227-7013 or 7014 (Feb 2014).
+//
+// This notice must appear in all copies of this file and its derivatives.
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcpputils/asserts.hpp"
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rcutils/filesystem.h"
+
+#include "rosbag2_cpp/logging.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/reindexer.hpp"
+
+#include "rosbag2_storage/storage_options.hpp"
+
+#ifdef WIN32
+// Import windows filesystem functionality
+#include <windows.h>
+#include <tchar.h>
+#include <stdio.h>
+#else
+// We're on a UNIX system. Import their filesystem stuff instead
+#include <dirent.h>
+#endif
+
+
+namespace rosbag2_cpp
+{
+Reindexer::Reindexer(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+: storage_factory_(std::move(storage_factory)),
+  metadata_io_(std::move(metadata_io))
+{
+  regex_bag_pattern_ = R"(.+_(\d+)\.([a-zA-Z0-9])+)";
+}
+
+Reindexer::~Reindexer() {}
```

Comment:
nit: you could leave this as `virtual ~Reindexer() = default` in the header if you wanted to.